### DISCUSSION TITLE
Could we get a basic example on how to convert reqif to excel / csv please?

### DISCUSSION THREAD

**Discussion by vesper8** [strictdoc-project/reqif#150] [Q&A]
Hello! First of all many thanks for making this project. I tried to implement a super basic reqif parser in Javascript and it worked for some very basic documents but now I'm running into issues with more complex documents so I've looked to your library which is clearly much much better than what I currently have. My current project requires that I be able to import requirements from reqif format and add them a simple relational database.

My first task in doing this would be to try and convert a reqif document to csv.

Your Readme.md states `The first-stage parser (implemented by the class ReqIFParser) can be used by user's second-stage parser/converter scripts that convert the ReqIF in-memory structure into a desired format such as Excel`

But I'm a little lost as to how to go about doing this. I was hoping there might be a simple example I could follow, or maybe just a code snippet that would show hot to extract requirement data from a reqif file and map it out to a simple array that I can then do as I wish with.

Any code snippet would be very appreciated to help me get started.

Many thanks!

**Comment by stanislaw** [comment]
Yeah, the documentation is very far from complete.

Could you try this and let me know if it is enough for you to get started? That code creates a single document with one requirement in it: https://github.com/strictdoc-project/reqif/blob/main/tests/integration/examples/01_create_reqif_objects/script.py.

**Reply by vesper8** [reply]
Thank you for linking to this example.. one of my client's planned requirements is the ability to convert requirements stored in a relational database back into the reqif format for the above example should prove to be quite useful for that! However I don't really how that can help me turn a .reqif file into csv or excel..

**Comment by stanislaw** [comment]
I could create an Excel example over this weekend but don't have it in hand right now.

**Reply by vesper8** [reply]
That would be incredibly awesome if you could do that. Reqif -> csv would be more than sufficient. Perhaps you could even improve your `dump` command to be able to accept an output format such as csv / excel in the future.

Basically I understand that what I need to accomplish is retrieving the column headers from attribute-definition-strings and then insert the matching attribute-value-xxx from the spec-objects and this way I can end up with a csv or excel format. But I'm just struggling on how best to accomplish this in a way that won't break with more complex reqif documents.

If you could find the time to create such an example over the weekend that would be incredibly helpful!!

Thank you very much!

**Reply by stanislaw** [reply]
I will try my best to do it this weekend 😄

**Reply by vesper8** [reply]
Did you by any chance find the time? : )

**Reply by stanislaw** [reply]
I could not finish what I wanted but only started it 😄 

This PR will track my work until it is done:

https://github.com/strictdoc-project/reqif/pull/151/files

Right now, the script is as easy as it can possibly be:

- It does not distinguish between requirements and chapters.
- It does not maintain levels like 1, 1.1, 1.2.3, etc.
- It does not recognize relations between requirements.
- It does not respect anything about the ReqIF file's type system.
- It has no idea of whether plain-text fields or rich XHTML fields are used.
- a few other points


Starting from this point, could you share which ReqIF files do you have in hand? Is there a chance you could send me your ReqIF file in an [anonymized form](https://github.com/strictdoc-project/reqif#anonymizing-reqif)?

**Reply by vesper8** [reply]
Thank you for working on it!! : )

Here is the file I have to work with right now, they haven't given me the full file yet just this sample. I've anonymized it manually but in any case it's test data.

I need to get as much data off of it as possible.. at the very least I need to get:

ChapterName
ForeignCreatedBy
ForeignCreatedOn
ForeignID
Text
Status

The output could be csv, but it could also be json.. json would be preferred in a way because that would allow it to include additional meta data such as data pertaining to the source reqif file at the time of importing the requirements. But certainly CSV would be a great start and very helpful.

Here is the file (I had to .zip it because Github doesn't like .reqif files):

[anon-import.reqif.zip](https://github.com/strictdoc-project/reqif/files/12841640/anon-import.reqif.zip)

Let me know if you need anything else.

Thanks a lot!

**Reply by stanislaw** [reply]
Just to make sure: this project is not my main job, so I might need one or more spare evenings to get the whole thing right. I will try to push the updates incrementally though, maybe something today.

**Reply by vesper8** [reply]
Totally understand! I will exercise patience and gratitude : )

**Reply by stanislaw** [reply]
Just a small heads up: I have been focusing on some other unresolved items, but this one is becoming the next one in my spare time work queue. I either get to it tomorrow or over the weekend.

**Reply by vesper8** [reply]
Cheers! Thanks for the update!

**Reply by stanislaw** [reply]
I have done the next step using your file, and now I am getting this CSV content:

```
Stanislaw@home reqif % cat output.csv
ReqIF.ChapterName,ReqIF.ForeignCreatedBy,ReqIF.ForeignCreatedOn,ReqIF.ForeignID,ReqIF.Text,ReqIF.ChapterName,ReqIF.ForeignCreatedBy,ReqIF.ForeignCreatedOn,ReqIF.ForeignID,ReqIF.Text,Status
SW: Lorem Ipsum,redacted@mail.com,2023-03-15T10:46:58.611Z,LOREM-818,"
                <xhtml:div>The Lorem Ipsum shall do something.</xhtml:div>
              ",SW: Lorem Ipsum,redacted@mail.com,2023-03-15T10:46:58.611Z,LOREM-818,"
                <xhtml:div>The Lorem Ipsum shall do something.</xhtml:div>
              ",Draft
```

This is how you can do it yourself:

```
pip install -U --pre git+https://github.com/strictdoc-project/reqif.git@stanislaw/examples
# Clone the reqif repository and from there:
reqif convert-to-csv tests/integration/csv/01_polarion_sample/sample.reqif .
```

The question is: is this what you are looking for? Or you want something more advanced? Please note that this implementation only works against the file you provided and may fail on more complex examples.

**Reply by vesper8** [reply]
Thanks so much man! Really appreciate it!

I tested it and it is workin! I have a bit of feedback if you don't mind

First of all, it seems that with the example you're using, it duplicates the columns and values? As you can see above, all columns are repeated a 2nd time and so are the values.

A few things that would make this absolutely perfect:

Remove 'ReqIF:' from the column names

Allow to name the output file ex:`reqif convert-to-csv sample.reqifz ./custom.csv`

Allow use with a reqifz file directly <3 <3

BONUS POINTS FOR:

Add json format

Command could then look like:

`reqif convert --format json --filename customName sample.reqifz .`

If this command worked I would be in heaven!

Here's what the json format could look like, with some provided flexibility for adding to the metadata field:

```
{
  "meta": {
    "filename": "sample.reqifz"
  },
  "columns": [
    "ChapterName",
    "ForeignCreatedBy",
    "ForeignCreatedOn",
    "ForeignID",
    "Text",
    "Status"
  ],
  "records": [
    {
      "ChapterName": "SW: Speed detection sensor",
      "ForeignCreatedBy": "soeren.woitack@moia.io",
      "ForeignCreatedOn": "2023-03-15T10:46:58.611Z",
      "ForeignID": "TRAIN-515",
      "Text": "<xhtml:div>The drive pilot shall have a sensor to detect the speed of other cars.</xhtml:div>",
      "Status": "Draft"
    }
  ]
}
```

Quite a few suggestions but I bet with your advanced knowledge of your tool you could have them all done in just a few minutes ;-D!!

Thanks!!!!!

**Reply by vesper8** [reply]
Actually.. one last item on my ultimate wishlist.. since I'll be calling your script through a server and then logging the output.. what would be really useful is if you could add one more optional flag for exporting directly to the CLI.. regardless of the export format.

Thank you!! You're awesome!!

**Reply by stanislaw** [reply]
Next iteration:

```json
{
    "documents": [
        {
            "name": "Test Export",
            "content": [
                {
                    "Type": "System Requirement",
                    "ReqIF.ForeignCreatedBy": "redacted@mail.com",
                    "ReqIF.ForeignID": "LOREM-818",
                    "Status": "Draft",
                    "ReqIF.ForeignCreatedOn": "2023-03-15T10:46:58.611Z",
                    "ReqIF.ChapterName": "SW: Lorem Ipsum",
                    "ReqIF.Text": "\n                <xhtml:div>The Lorem Ipsum shall do something.</xhtml:div>\n              "
                }
            ]
        }
    ],
    "fields": [
        "ReqIF.ChapterName",
        "ReqIF.ForeignCreatedBy",
        "ReqIF.ForeignCreatedOn",
        "ReqIF.ForeignID",
        "ReqIF.Text",
        "Status"
    ]
}
```

The reason the fields seemed to be duplicated was because your reqif has two types: one for a heading and one for a requirement. Note that your example with just one spec object is rather basic. Could you provide one that has one or more sections/chapters with requirements nested in them?

Now that we are going to model most of ReqIF's complexity in JSON, and the CSV definitely doesn't scale here, we also need to model the `fields` for each type (heading or requirement). Also, in order to be able to recognize how nested your requirements or sections are, I need to introduce the level field, like 1.1, 1.2.3, etc. because otherwise you will not know who contains what. Either the list of objects is kept plain with the `Level: ` field or the JSON has to nest the content with sections nesting requirements or other sections.

To summarize: it is not a very quick task if the converter has to deal with documents of arbitrary structure. I was going to implement all of this anyway, it is just that a good and robust implementation might take time. 

I will get to all your comments, especially to those about the CLI design, at my next pass to this project somewhere next week. For now, the CLI is the same (still csv in it).

Also: Is this important for you to be able to convert JSON back to ReqIF? If yes, then it is especially import that the JSON representation preserves enough data and other element types information.

**Reply by vesper8** [reply]
Very excited for your upcoming iteration and thank you very much for taking all my comments into consideration!

I have asked for more complex reqif examples but am unsure if they will reliably come in anytime soon... in the meantime I have found these reqif samples that appear to be of varying levels of complexity. I found this in your own Github issues. Finding complex reqif samples on the internet is not an easy task..

https://ci.eclipse.org/rmf/job/rmf.develop.mars/lastSuccessfulBuild/artifact/test-data/

Although I don't know if any of them include chapters and links. In any case, for my current use case I would not need to import requirement linkage and being able to import a flattened array of requirements would suffice for now.

**Reply by vesper8** [reply]
Hello my friend! Just checking in to see if you've had time to work on the CLI improvements and other feedback I gave above or if you anticipate you'll have time soon? Thank you!

**Reply by stanislaw** [reply]
Hey @vesper8, not yet. The higher priority parent project's release triggered some hotfixes and the related discussion. I was planning to get to this JSON branch this evening.
